File: Code/Entities/Celeste/ArrowHole.cs

using System;
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.XaphanHelper.Entities
{
    [Tracked(true)]
    [CustomEntity("XaphanHelper/ArrowHole")]
    class ArrowHole : Entity
    {
        private string side;

        private Sprite sprite;

        private string directory;

        public ArrowHole(EntityData data, Vector2 offset) : base(data.Position + offset)
        {
            side = data.Attr("side", "Left");
            directory = data.Attr("directory", "objects/XaphanHelper/ArrowHole");
            Add(sprite = new Sprite(GFX.Game, directory + "/"));
            sprite.Add("hole", "hole", 0f);
            sprite.Origin = new Vector2(sprite.Width / 2, sprite.Height / 2);
            sprite.Play("hole");
            if (side == "Left")
            {
                Collider = new Hitbox(1, 2, 0, 7);
                sprite.Position = new Vector2(4f, 8f);
                sprite.Rotation = -(float)Math.PI / 2f;
            }
            else if (side == "Right")
            {
                Collider = new Hitbox(1, 2, 7, 7);
                sprite.Position = new Vector2(4f, 8f);
                sprite.Rotation = (float)Math.PI / 2f;
            }
            else if (side == "Top")
            {
                Collider = new Hitbox(2, 1, 7, 7);
                sprite.Position = new Vector2(8f, 4f);
                sprite.Rotation = (float)Math.PI;
            }
            else if (side == "Bottom")
            {
                Collider = new Hitbox(2, 1, 7, 0);
                sprite.Position = new Vector2(8f, 4f);
            }
            Depth = -15000;
        }

        public override void Render()
        {
            base.Render();
        }
    }
}

File: Loenn/entities/ArrowHole.lua
local drawableSprite = require("structs.drawable_sprite")

local ArrowHole = {}

ArrowHole.name = "XaphanHelper/ArrowHole"
ArrowHole.depth = -15000
ArrowHole.fieldInformation = {
    side = {
        options = {"Left", "Right", "Top", "Bottom"},
        editable = false
    }
}
ArrowHole.placements = {
    name = "ArrowHole",
    data = {
        side = "Left",
        directory = "objects/XaphanHelper/ArrowHole"
    }
}

function ArrowHole.sprite(room, entity)
    local texture = entity.directory or "objects/XaphanHelper/ArrowHole"
    local side = entity.side or "Right"

    local sprite = drawableSprite.fromTexture(texture .. "/hole00", entity)

    if side == "Right" then
        sprite.rotation = math.pi / 2
    elseif side == "Left" then
        sprite.rotation = -math.pi / 2
    elseif side == "Top" then
        sprite.rotation = math.pi
        sprite:addPosition(4, -4)
    elseif side == "Bottom" then
        sprite:addPosition(4, -4)
    end
    sprite:addPosition(4, 8)

    return sprite
end

return ArrowHole